File: C++/1237.cpp
__________________________________________________________________________________________________
sample 0 ms submission
/*
 * // This is the custom function interface.
 * // You should not implement it, or speculate about its implementation
 * class CustomFunction {
 * public:
 *     // Returns f(x, y) for any given positive integers x and y.
 *     // Note that f(x, y) is increasing with respect to both x and y.
 *     // i.e. f(x, y) < f(x + 1, y), f(x, y) < f(x, y + 1)
 *     int f(int x, int y);
 * };
 */
/*
- - - + + + +
- - - + + + +
- - - z + + +
- - - - - - -
- - - - - - -

*/
        //x,y++排除
        //x++,y++排除
        //x++,y排除
//
class Solution {
public:
    vector<vector<int>> findSolution(CustomFunction& customfunction, int z) {
    vector<vector<int>> re;
        int middle;
    for(int x=1;x<1001;x++)
    {
        int left=1,right=1000;
        while(left<=right)
        {
            middle=(left+right)/2;
            if(customfunction.f(x,middle)<z)
                left=(left+right)/2+1;
            else if(customfunction.f(x,middle)>z)
                right=middle-1;
            else 
            {          
                re.push_back(vector<int>{x,middle});
                break;
            }

        }
    }
        return re;
    return re;
    }
};
__________________________________________________________________________________________________
sample 4 ms submission
/*
 * // This is the custom function interface.
 * // You should not implement it, or speculate about its implementation
 * class CustomFunction {
 * public:
 *     // Returns f(x, y) for any given positive integers x and y.
 *     // Note that f(x, y) is increasing with respect to both x and y.
 *     // i.e. f(x, y) < f(x + 1, y), f(x, y) < f(x, y + 1)
 *     int f(int x, int y);
 * };
 */

class Solution {
public:
    vector<vector<int>> findSolution(CustomFunction& customfunction, int z) {
        
        vector<vector<int>> ans;
        
        for(int i = 1; i <= 1000; i++) {
            for(int j = 1; j <= 1000; j++) {
            
                if(customfunction.f(i, j) == z) {
                    
                    vector<int> tmp(2);
                    tmp[0] = i;
                    tmp[1] = j;
                    ans.push_back(tmp);
                } else if(customfunction.f(i, j) >= z)
                    break;
                
            }
        }
        return ans;
    }
};
__________________________________________________________________________________________________

File: Python3/1237.py
__________________________________________________________________________________________________
sample 16 ms submission
"""
   This is the custom function interface.
   You should not implement it, or speculate about its implementation
   class CustomFunction:
       # Returns f(x, y) for any given positive integers x and y.
       # Note that f(x, y) is increasing with respect to both x and y.
       # i.e. f(x, y) < f(x + 1, y), f(x, y) < f(x, y + 1)
       def f(self, x, y):
  
"""
class Solution:
    def findSolution(self, customfunction: 'CustomFunction', z: int) -> List[List[int]]:
        out_list = []
        x = 1
        while x<1000:
            y = 1
            if customfunction.f(x, y) > z:
                break
            while y<1000:
                if customfunction.f(x, y) > z:
                    break
                if customfunction.f(x, y) == z:
                    out_list.append([x,y])
                y+=1
            x+=1
        return out_list
__________________________________________________________________________________________________
sample 20 ms submission
"""
   This is the custom function interface.
   You should not implement it, or speculate about its implementation
   class CustomFunction:
       # Returns f(x, y) for any given positive integers x and y.
       # Note that f(x, y) is increasing with respect to both x and y.
       # i.e. f(x, y) < f(x + 1, y), f(x, y) < f(x, y + 1)
       def f(self, x, y):
  
"""
class Solution:
    def findSolution(self, customfunction: 'CustomFunction', z: int) -> List[List[int]]:
        def f(x, y):
            return customfunction.f(x, y)
            
        def find_sol(y, x_L, x_R):
            while x_L<=x_R:
                mid = x_L+(x_R-x_L)//2
                f_mid = f(mid, y)
                if(f_mid == z):
                    return [mid, y]
                elif(f_mid < z):
                    x_L = mid+1
                else:
                    x_R = mid-1
            return []
            
        #print (f(1, 1), f(4, 1))
        #return
    
        if(f(1, 1)>z):
            return []
        elif(f(1, 1)==z):
            return [[1, 1]]
        x_R = 1
        while(f(x_R, 1)<z):
            x_R *= 2
        sol = []
        y = 1
        while x_R>=1 and f(1, y)<=z:
            #print (y, x_R)
            tmp = find_sol(y, 1, x_R)
            if(tmp):
                sol.append(tmp)
            y += 1
            if(tmp):
                x_R = tmp[0]-1
        return sol
__________________________________________________________________________________________________
